Fix #4236 - Use SceneDelegate instead of AppDelegate

Brandon-T opened the pull request:
## Summary of Changes

- Change from AppDelegate to properly support SceneDelegate.
- This will bring us closer to supporting multiple windows and getting rid of all the singletons everywhere.
- Fixes the issues CarPlay introduced by using SceneDelegate for iOS 14+.
- Should have moved a long time ago :)

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #4236, fixes #4228

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
- Test opening URLs from other apps (Safari, Twitter, etc)
- Test Sharing URLs
- Test Playlist (Pip, Carplay, Error messages/popups)
- Test Opening and Closing Tabs (NTP + Regular browsing)
- Test Rewards
- Test Stats on NTP
- Test Onboarding + Selecting search engine from Onboarding
- Test Sync
- Test ReaderMode
- Test Private Browsing


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

iccub reviewed (changes requested):
Looks super clean and a lot of code comments which i like,
will run this app now to catch any early bugs

iccub reviewed (approved):
Good job on this unexpected task :)
I ran the app and it seems to work well

Brandon-T committed:
Address feedback.

kylehickinson reviewed (approved):
🎉

Brandon-T committed:
Use SceneDelegate instead of AppDelegate.

Fix Unit Tests

Pass in profile where we can.

iOS 14+ updates.

TrackingProtectionPageStats should pass in the current tab instead of accessing the selected tab via AppDelegate.
